Make the container fully fulfill PSR container interface

susnux opened the pull request:
## Summary
To fulfill the PSR container interface it must throw `NotFoundExceptionInterface` if the requested name was not found within the container. 
Added a simple private wrapper over the `QueryException` implementing the `NotFoundExceptionInterface` which is only thrown if the query was not found, this way the public API should be unchanged and only the thrown error **additionally** implements `NotFoundExceptionInterface`.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

susnux committed:
fix: Throw `NotFoundExceptionInterface` to fulfill PSR container interface if class not found

Signed-off-by: Ferdinand Thiessen <rpm@fthiessen.de>

nickvergessen reviewed (approved):
https://www.php-fig.org/psr/psr-11/#33-psrcontainernotfoundexceptioninterface

ChristophWurst reviewed (approved):
Thanks!

ChristophWurst commented:
https://github.com/nextcloud/3rdparty/blob/1d53ed4d3282427854fca7ee6ecbb945304272db/composer.json#L47 time to bump that to v2

susnux commented:
> time to bump that to v2

Would include extending `Throwable` to the exceptions and adding the boolean return type hint to the `has` method.
As far as I can see this is already done on our implementation (extending `Exception` which implements `Throwable`), so this should be no breaking change.
Should I include this within this PR?

---

ChristophWurst commented:
Let's only fix the PSR compliance here to keep this backportable and do the bump for master/26 only.

susnux commented:
I have no clue why drone is failing, but it seems unrelated

ChristophWurst commented:
Test\AppFramework\Utility\SimpleContainerTest::testNothingRegistered sounds related

ChristophWurst commented on tests/lib/AppFramework/Utility/SimpleContainerTest.php:
```diff
@@ -78,9 +79,13 @@ public function testRegister() {
+			$this->container->query('something really hard');
+			$this->fail('Expected `QueryException` exception was not thrown');
+		} catch (Exception $exception) {
+			$this->assertTrue($exception instanceof \OCP\AppFramework\QueryException);
```
nit: ``self::assertInstanceOf(\OCP\AppFramework\QueryException::class, $exception)``

susnux commented on lib/public/AppFramework/QueryException.php:
```diff
@@ -32,7 +32,7 @@
  * Class QueryException
  *
- * The class extends `NotFoundExceptionInterface` since 20.0.0
+ * The class extends `ContainerExceptionInterface` since 20.0.0
```
Noticed this commend was wrong, as the class did not extend the not found interface but the container exception interface. Fixed.

susnux commented:
I fixed that issue, but that CI still fails with some issues on the encryption, which seem definitely unrelated, but I rebased onto current master to hopefully fix that drone nodb  check.

susnux commented:
Well the remaining drone issues seem to be unrelated, as the same issues are reported for every PR currently ...

nickvergessen commented:
> There was 1 failure:
> 1) Test\AppFramework\Utility\SimpleContainerTest::testNothingRegistered
> Failed asserting 
that OCP\AppFramework\QueryException Object (...) is an instance of 
interface "Psr\Container\NotFoundExceptionInterface".
> 
> /drone/src/tests/lib/AppFramework/Utility/SimpleContainerTest.php:87

nickvergessen commented:
For the encryption error just rebase, the above error is your code change and needs fixing/adjusting

susnux commented:
@nickvergessen thank you! I forgot to handle querying with autoloading enabled, this is fixed now and the test passes.

ChristophWurst commented on lib/composer/composer/autoload_classmap.php:
```diff
@@ -720,6 +720,7 @@
     'OC\\AppFramework\\Services\\AppConfig' => $baseDir . '/lib/private/AppFramework/Services/AppConfig.php',
     'OC\\AppFramework\\Services\\InitialState' => $baseDir . '/lib/private/AppFramework/Services/InitialState.php',
     'OC\\AppFramework\\Utility\\ControllerMethodReflector' => $baseDir . '/lib/private/AppFramework/Utility/ControllerMethodReflector.php',
+    'OC\\AppFramework\\Utility\\QueryNotFoundException' => $baseDir . '/lib/private/AppFramework/Utility/QueryNotFoundException.php',
```
nit: this is the only change in the autoloader files you need to commit

susnux replied:
There was a drone error:
```
The autoloaders are not up to date
Please run: bash build/autoloaderchecker.sh
```
Just committed all changed files, but I can undo this

susnux committed:
Update autoloaders

Signed-off-by: Ferdinand Thiessen <rpm@fthiessen.de>

nickvergessen commented:
Something to document in https://github.com/nextcloud/server/issues/34692 ?

susnux commented:
/backport to stable25
